### DISCUSSION TITLE
ORM & Database

### DISCUSSION THREAD

**Discussion by reindert-vetter** [confetti-framework/confetti#110] [Ideas]
I've thought about that for a long time. But I think we should wait for Generics in Go. I can make all subjects in such a way that gives the same feeling as Laravel, only an ORM needs generics in my opinion.

We could create or accept a query builder. How do we think about that?

**Comment by reindert-vetter** [comment]
A list of ORMs:
https://github.com/d-tsuji/awesome-go-orms

**Comment by rafaelbreno** [comment]
I'm used to [GORM](https://github.com/go-gorm/gorm) but I feel a little lack of a well written documentation, looks like I'm reading the code comments instead of the official doc.

Seizing the theme, I'm assigned to the [Validation Depth](https://github.com/confetti-framework/validation/issues/1) and an ORM would be required for some validation, for example the "unique" keys, would need to verify the table and search for a value(you know). So if there's no rush to implement an ORM would be nice to wait for a Generics so we could think and project calmly an Eloquent-ish package.

**Comment by Wulfheart** [comment]
I like the idea of waiting.

**Comment by bosix** [comment]
I use gorm in some of my projects as well and building the query feels really a bit of "laravel-like". Here an example out of one of my projects:

```golang
dbc := db.Connection()
courseDataReversed := []db.HistoricCourse{}
dbc.Select("avg(btc) as btc, avg(eth) as eth, avg(ltc) as ltc, avg(bch) as bch, min(id) as id, created_at").
	Table("historic_courses").
	Where("created_at < ?", t.currentTime).
	Where("created_at > ?", t.currentTime.Add((-1)*24*8*time.Hour)).
	Group("created_at_quarter_hour").
	Order("created_at DESC").
	Limit(7*24*4 + 5).
	Find(&courseDataReversed)
```

The syntax is not that beautiful than it's in laravel but I feel that it's kinda easy to use. I have to admit that @rafaelbreno is right according the documentation. If you have special things to solve (complex subqueries, multiple joins, etc) there are very few examples to work with.

**Reply by Wulfheart** [reply]
But gorm doesn’t have the simplicity like Eloquent.

**Reply by bosix** [reply]
@Wulfheart can you give an example of what you mean?

**Reply by Wulfheart** [reply]
Wait, I mixed it up with something else. 
However, I suggest using gorm just until we can implement a solution using generics. But there really needs to be a default solution. Even though this is temporary.

**Reply by reindert-vetter** [reply]
> But there really needs to be a default solution. Even though this is temporary.

Do you want to explain way?

**Reply by Wulfheart** [reply]
@reindert-vetter just for the sake of simplicity. Same defaults…

**Comment by almas-x** [comment]
Recommend  [REL ](https://go-rel.github.io/)

> REL is golang orm-ish database layer for layered architecture. It's testable and comes with it's own test library. REL also features extendable query builder that allows you to write query using builder or plain sql.

**Reply by reindert-vetter** [reply]
That looks good indeed!

**Reply by Wulfheart** [reply]
I'd opt for replicating Eloquent in Go with the advent of generics.

**Comment by nurmuhammet-ali** [comment]
Hello everyone, I really loved the framework, but I don't where it will be better to put database functions, I mean I used to open one database connection in my main.go file then I use it any where, just by passing credential. So, the question is, where it will be better to put my db functions, what I used to do is:
```go
// In main.go
db, err := sql.Open("mysql", dbUsername + ":" + dbPassword + "@/" + dbName)

if err != nil {
 panic(err.Error())
}
defer db.Close()

// then in my other functions/routes, I pass the db, then credentials
createData(db, ...)

// here is the function
func createData(db *sql.DB, ...) {/* ... */}
```

Thx for attention.

**Reply by reindert-vetter** [reply]
@nurmuhammet-ali I think it is best to do this in a service provider. However, you cannot execute `defer db.Close()` there. You're using https://golang.org/pkg/database/sql/, right? I think I'm going to work out a solution for this tonight or tomorrow (A stable implementation with documentation can still take a while.). I will use a service provider and ultimately, the configuration comes in a config file. If I implement that, we will have a nice solution that can stay in the framework. Later we could implement an ORM.

Do you have any ideas and tips that I should pay attention to?

**Reply by nurmuhammet-ali** [reply]
@reindert-vetter Great, thx. It would be nice if we could perform database operations after defining the database type and configuring the credentials on env file.

**Reply by nurmuhammet-ali** [reply]
@reindert-vetter For example,
```go
/* In Controllers */
// By dependency injection
func Homepage(request inter.Request, db *sql.DB) inter.Response {
	table, _ := db.Query("select ... from table")
       /*  ...  */
}
```

What do you think? Is it appropriate?

**Reply by reindert-vetter** [reply]
Good to know https://golang.org/pkg/database/sql/#Open:
> The returned DB is safe for concurrent use by multiple goroutines and maintains its own pool of idle connections. Thus, the Open function should be called just once. It is rarely necessary to close a DB.

**Reply by reindert-vetter** [reply]
I think you will indeed get something like this. Maybe you should resolve it from the app or pass the app along.

**Reply by reindert-vetter** [reply]
Sneek preview: 👀 
``` go
package config

import (
	"github.com/confetti-framework/contract/inter"
	"github.com/confetti-framework/foundation/db"
	"github.com/confetti-framework/support/env"
	"time"
)

var Database = struct {
	Default     string
	Connections map[string]inter.Connection
}{

	/*
	   |--------------------------------------------------------------------------
	   | Default Database Connection Name
	   |--------------------------------------------------------------------------
	   |
	   | Here you may specify which of the database connections below you wish
	   | to use as your default connection for all database work. Of course
	   | you may use many connections at once.
	   |
	*/
	Default: env.StringOr("DB_CONNECTION", "mysql"),

	/*
	   |--------------------------------------------------------------------------
	   | Database Connections
	   |--------------------------------------------------------------------------
	   |
	   | Here are each of the database connections setup for your application.
	   | Of course, examples of configuring each database platform that is
	   | supported by Confetti is shown below to make development simple.
	   |
	*/

	Connections: map[string]inter.Connection{
		"mysql": db.MySQL{
			Host:         env.StringOr("DB_HOST", "127.0.0.1"),
			Port:         env.IntOr("DB_PORT", 3306),
			Database:     env.StringOr("DB_DATABASE", "confetti"),
			Username:     env.StringOr("DB_USERNAME", "app"),
			Password:     env.StringOr("DB_PASSWORD", ""),
			QueryTimeout: 10 * time.Second,
		},
		"postgresql": db.PostgreSQL{
			Host:         env.StringOr("DB_HOST", "127.0.0.1"),
			Port:         env.IntOr("DB_PORT", 5432),
			Database:     env.StringOr("DB_DATABASE", "confetti"),
			Username:     env.StringOr("DB_USERNAME", "app"),
			Password:     env.StringOr("DB_PASSWORD", ""),
			QueryTimeout: 10 * time.Second,
		},
	},
}
```

**Reply by reindert-vetter** [reply]
The syntax will look like this:

INSERT and UPDATE:
``` go
result := app.Db().Exec("INSERT INTO employee (first_name, last_name) VALUES ('Reindert','Vetter')")
```
Or:
``` go
result, err := app.Db().ExecE("INSERT INTO employee (first_name, last_name) VALUES ('Reindert','Vetter')")
```

SELECT:
```go
result := app.Db().Query("SELECT first_name FROM employee;")
```
Or:
``` go
result, err := app.Db().QueryE("SELECT first_name FROM employee;")
```

In a controller it could look like this:
``` go
func Homepage(request inter.Request) inter.Response {
    db := request.App().Db()
    result := db.Query("SELECT first_name FROM employee;")
    if result.Empty() {
        return outcome.Json(errors.New("Not found").Status(http.StatusNotFound))
    }
    return outcome.Html("Name: " + result.Get("0.employee_first_name").String())
}
```

Or
``` go
func Homepage(request inter.Request) inter.Response {
    db := request.App().Db()
    amount := db.Query("SELECT count(*) FROM employee;").First().Int()
    return outcome.Html("Amount: " + amount)
}
```

I have now only made it for MySQL, but can also make it easy for PostgreSQL. 
I will document it. Then it will become clearer how you can get values from the result.

@nurmuhammet-ali Do you already have points for improvement?

**Reply by nurmuhammet-ali** [reply]
@reindert-vetter Great, it would be perfect, that's exactly what I wanted.

**Reply by reindert-vetter** [reply]
With arguments:
``` go
	result := app.Db().Query(
		"SELECT first_name "+
			"FROM employee "+
			"WHERE first_name = ? "+
			"AND last_name = ?;",
		"Reindert",
		"Vetter",
	)
```

**Reply by reindert-vetter** [reply]
@nurmuhammet-ali 

I'm ready. The following has been changed in confetti-framework/confetti repo:
PR: https://github.com/confetti-framework/confetti/pull/132

I just need to make sure it works in docker. Now I get error:
```
panic: can't connect with connection mysql: Error 1045: Access denied for user 'app'@'172.19.0.5' (using password: YES)
```

**Comment by lightglitch** [comment]
Also have https://entgo.io

**Reply by reindert-vetter** [reply]
Love it

**Comment by Wulfheart** [comment]
Generics are available now. Are there any plans?

**Comment by emarais-godaddy** [comment]
Our team has implemented https://github.com/volatiletech/sqlboiler successfully in some projects before. It's a great way to setup models from your migrations and then use those models those setup database seeds.

It takes away the heavy lifting of creating db boilerplate throughout projects.

**Reply by nurmuhammet-ali** [reply]
Looks really nice!
